### PR TITLE
MCPayment: add ERC-20 payment

### PR DESCRIPTION
**Pull request by volodymyr-basiuk**

### CODE REVIEW

File: contracts/payment/MCPayment.sol [hunk 1]
```diff
@@ -18,6 +19,12 @@ contract MCPayment is Ownable2StepUpgradeable, EIP712Upgradeable {
             "Iden3PaymentRailsRequestV1(address recipient,uint256 amount,uint256 expirationDate,uint256 nonce,bytes metadata)"
         );
 
+    bytes32 public constant ERC_20_PAYMENT_DATA_TYPE_HASH =
+        keccak256(
+            // solhint-disable-next-line max-line-length
+            "Iden3PaymentRailsERC20RequestV1(address tokenAddress,address recipient,uint256 amount,uint256 expirationDate,uint256 nonce,bytes metadata)"
+        );
+
```
Comment (daveroga) [inline contracts/payment/MCPayment.sol:27]: Maybe we could have only 1 PAYMENT_DATA_TYPE adding the `tokenAddress` and check that this `tokenAddress` is 0x0 for native token?

Comment (volodymyr-basiuk) [reply contracts/payment/MCPayment.sol:27]: Not sure that validating in each native token payment is better than having one more constant. @AndriianChestnykh what do you think ?

Comment (AndriianChestnykh) [reply contracts/payment/MCPayment.sol:27]: @volodymyr-basiuk I agree. Different reasons to change
You should do `erc20Payment` payable then and check msg.value == 0 or !=0 in different cases, can't securely resent ETH, but may resend tokens to issuer... what if you add other than erc20 asset payment, etc..
However two function names `pay` and `erc20Payment` could've been more consistent to each other.

File: contracts/payment/MCPayment.sol [hunk 2]
```diff
@@ -143,6 +194,29 @@ contract MCPayment is Ownable2StepUpgradeable, EIP712Upgradeable {
...
+                paymentData.nonce,
+                keccak256(paymentData.metadata)
+            )
+        );
+        bytes32 hashTypedData = _hashTypedDataV4(structHash);
+        (address recovered, ECDSA.RecoverError err, ) = hashTypedData.tryRecover(signature);
+
+        if (err != ECDSA.RecoverError.NoError || recovered != paymentData.recipient) {
+            revert InvalidSignature("MCPayment: invalid signature");
+        }
```
Comment (AndriianChestnykh) [inline contracts/payment/MCPayment.sol:217]: Repeated part?
Also worth using different error or error message?

File: contracts/payment/MCPayment.sol (hunk 1 above)
Comment (volodymyr-basiuk) [reply contracts/payment/MCPayment.sol:27]: I’m not in favor of using a single hash, as these hashes represent two different payment types at the protocol level. We previously discussed reusing the Iden3PaymentRailsRequestV1 type for ERC-20 payments and, if the payment was made on the native currency, passing 0x for the tokenAddress. However, this approach only adds confusion, so we decided to provide a separate type - Iden3PaymentRailsERC20RequestV1.

File: contracts/payment/MCPayment.sol [hunk 3]
```diff
@@ -135,12 +186,47 @@ contract MCPayment is Ownable2StepUpgradeable, EIP712Upgradeable {
...
+        );
+
+        if (!isSignatureValid(structHash, signature, paymentData.recipient)) {
+            revert InvalidSignature(
+                "MCPayment: invalid signature for Iden3PaymentRailsERC20RequestV1"
+            );
+        }
+    }
+
+    function isSignatureValid(
```
Comment (AndriianChestnykh) [inline contracts/payment/MCPayment.sol:217]: better to use "_" in the beginning of all internal and private functions. Easier to real function lists.

File: contracts/payment/MCPayment.sol [hunk 4]
```diff
@@ -89,21 +106,33 @@ contract MCPayment is Ownable2StepUpgradeable, EIP712Upgradeable {
         $.ownerPercentage = ownerPercentage;
     }
 
-    function pay(
+    function nativeCurrencyPayment(
```
Comment (AndriianChestnykh) [inline contracts/payment/MCPayment.sol:109]: `payNativeCurrency` and `payERC20Token` may be somewhat better

Comment (volodymyr-basiuk) [reply contracts/payment/MCPayment.sol]: sure, renamed

**Review by AndriianChestnykh — APPROVED** [review]

**Review by daveroga — DISMISSED** [review]

File: contracts/payment/MCPayment.sol (hunk 1 above)
Comment (vmidyllic) [reply contracts/payment/MCPayment.sol:27]: two different data domains for different protocol types is made decision for now.

File: contracts/payment/MCPayment.sol (hunk 4 above)
Comment (vmidyllic) [reply contracts/payment/MCPayment.sol]: vote just for 
- pay ( we omit word native if it is native in protocol)
- payERC20  (omit word token)

Comment (daveroga) [reply contracts/payment/MCPayment.sol]: Yes, I agree with @vmidyllic proposal for payment functions names. They look simpler and more clear.

Comment (volodymyr-basiuk) [reply contracts/payment/MCPayment.sol]: renamed

**Review by daveroga — DISMISSED** [review]

**Comment by volodymyr-basiuk** [comment]
Added new type - Iden3PaymentRailsEIP2612RequestV1

File: contracts/payment/MCPayment.sol [hunk 5]
```diff
@@ -116,12 +147,99 @@ contract MCPayment is Ownable2StepUpgradeable, EIP712Upgradeable {
...
+        verifyIden3PaymentRailsERC20RequestV1Signature(paymentData, signature);
+        bytes32 paymentId = keccak256(abi.encode(paymentData.recipient, paymentData.nonce));
+        MCPaymentStorage storage $ = _getMCPaymentStorage();
+        if ($.isPaid[paymentId]) {
+            revert PaymentError(
+                paymentData.recipient,
+                paymentData.nonce,
+                "MCPayment: payment already paid"
+            );
+        }
```
Comment (daveroga) [inline contracts/payment/MCPayment.sol:194]: This code is repeated in both `payERC20` and `payERC20Permit`. Maybe we could create a function `checkERC20Payment(paymentData, signature)` or something similar with the code.

**Review by daveroga — APPROVED** [review]

**Review by AndriianChestnykh — APPROVED** [review]

**Review by vmidyllic — APPROVED** [review]

**Pull request merged**